**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №8

Набор данных: credit_risk.csv

Атрибуты: \
Id — идентификатор \
Age — возраст \
Income — доход \
Home — статус домовладения (Собственный, Ипотека, Аренда) \
Emp_length — стаж работы в годах \
Intent — цель кредита \ 
Amount — сумма кредита по заявке \ 
Rate — процентная ставка \
Status — статус одобрения кредита \
Percent_income — сумма кредита в процентах от дохода \
Default — не допустил ли заявитель ранее дефолта по кредиту (да, нет) \
Cred_length — длина кредитной истории заявителя

### 1.Чтение файла (набора данных)

In [1]:
import pandas as pd

file = 'credit_risk.csv'
df = pd.read_csv(file, sep=',')

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [2]:
# применить метод head
df.head(20)


Id   Age  Income      Home  Emp_length             Intent  Amount   Rate  \
0    0  22.0   59000      RENT       123.0           PERSONAL   35000  16.02   
1    1  21.0    9600       OWN         5.0          EDUCATION    1000  11.14   
2    2  25.0    9600  MORTGAGE         1.0            MEDICAL    5500  12.87   
3    3  23.0   65500      RENT         4.0            MEDICAL   35000  15.23   
4    4  24.0   54400      RENT         8.0            MEDICAL   35000  14.27   
5    5  21.0    9900       OWN         2.0            VENTURE    2500   7.14   
6    6  26.0   77100     RENET         8.0          EDUCATION   35000  12.42   
7    7  24.0   78956      RENT         5.0            MEDICAL   35000  11.11   
8    8  24.0   83000      RENT         8.0           PERSONAL   35000   8.90   
9    9  21.0   10000       OWN         6.0            VENTURE    1600  14.74   
10  10  22.0   85000      RENT         6.0            VENTURE   35000  10.37   
11  11  21.0   10000       OWN         2.0    HOMEIMPROVEMENT    4500   8.63   
12  12  23.0   95000      RENT         2.0            VENTURE   35000   7.90   
13  13  26.0  108160      RENT         4.0          EDUCATION   35000  18.39   
14  14  23.0  115000      RENT         2.0          EDUCATION   35000   7.90   
15  15  23.0  500000  MORTGAGE         7.0  DEBTCONSOLIDATION   30000  10.65   
16  16  23.0  120000      RENT         0.0          EDUCATION   35000   7.90   
17  17  23.0   92111      RENT         7.0            MEDICAL   35000  20.25   
18  18  23.0  113000      RENT         8.0  DEBTCONSOLIDATION   35000  18.25   
19  19  24.0   10800  MORTGAGE         8.0          EDUCATION    1750  10.99   

    Status  Percent_income Default  Cred_length  
0        1            0.59       Y            3  
1        0            0.10       N            2  
2        1            0.57       N            3  
3        1            0.53       N            2  
4        1            0.55       Y            4  
5        1            0.25       N            2  
6        1            0.45       N            3  
7        1            0.44       N            4  
8        1            0.42       N            2  
9        1            0.16       N            3  
10       1            0.41       N            4  
11       1            0.45       N            2  
12       1            0.37       N            2  
13       1            0.32       N            4  
14       0            0.30       N            4  
15       0            0.06       N            3  
16       0            0.29       N            4  
17       1            0.32       N            4  
18       1            0.31       N            4  
19       1            0.16       N            2

2.2 Оценка данных с помощью метода info.

In [3]:
# выполнит метод info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 652 entries, 0 to 651
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Id              652 non-null    int64  
 1   Age             652 non-null    float64
 2   Income          652 non-null    int64  
 3   Home            652 non-null    object 
 4   Emp_length      643 non-null    float64
 5   Intent          652 non-null    object 
 6   Amount          652 non-null    int64  
 7   Rate            586 non-null    float64
 8   Status          652 non-null    int64  
 9   Percent_income  652 non-null    float64
 10  Default         652 non-null    object 
 11  Cred_length     652 non-null    int64  
dtypes: float64(4), int64(5), object(3)
memory usage: 61.3+ KB


2.3 Оценка данных с помощью метода describe.

In [4]:
# оцените числовые столбцы с помощью describe
df.describe()

Id         Age         Income  Emp_length        Amount  \
count  652.000000  652.000000     652.000000  643.000000    652.000000   
mean   325.495399   24.285276   90008.855828    4.601866  18800.996933   
std    188.352535    7.841061   69530.312525    7.281616   9152.896250   
min      0.000000   21.000000    9600.000000    0.000000   1000.000000   
25%    162.750000   23.000000   44000.000000    2.000000  10000.000000   
50%    325.500000   24.000000   69998.000000    4.000000  21850.000000   
75%    488.250000   25.000000  128499.000000    7.000000  25000.000000   
max    649.000000  144.000000  500000.000000  123.000000  35000.000000   

             Rate      Status  Percent_income  Cred_length  
count  586.000000  652.000000      652.000000   652.000000  
mean    12.289471    0.601227        0.279831     3.012270  
std      3.272772    0.490022        0.150231     0.808843  
min      5.420000    0.000000        0.010000     2.000000  
25%     10.250000    0.000000        0.160000     2.000000  
50%     12.180000    1.000000        0.280000     3.000000  
75%     14.720000    1.000000        0.380000     4.000000  
max     21.210000    1.000000        0.830000     4.000000


---

**Выводы**:
- Id - кол-во non-null 652 значений, а max id - 649, значит какой-то инремент id пропущен
- Age - максимальный возраст - 144 года, что не ествественно
- Income - средний доход около 90000
- Emp_length - 50 персинтиль / медиана стажа 4, а макс. стаж - 123, что не ествественно
- Amount - сумма кредита варируется от 1000 до 35000
- Rate - 75% клиентов взяли кредит по ставке меньше 14,72%
- Status - 652 строки имеют non-null статус
- Percent_income - макс. сумма кредита в процентах от дохода составила 0,83
- Cred_lenght - в среднем длина кредитной истории клиента равна 3

 ---


 2.4 Оценка названий столбцов

In [5]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.
df.columns

Index(['Id', 'Age', 'Income', 'Home', 'Emp_length', 'Intent', 'Amount', 'Rate',
       'Status', 'Percent_income', 'Default', 'Cred_length'],
      dtype='object')

In [6]:
# переименование при необходимости
df.columns = df.columns.str.lower()
df.columns

Index(['id', 'age', 'income', 'home', 'emp_length', 'intent', 'amount', 'rate',
       'status', 'percent_income', 'default', 'cred_length'],
      dtype='object')

**Пояснения**:
с неймингом проблем не было, но хорошей практикой считается наименование в lower кейсе, поэтому названия были приведены к соотвествующему кейсу

### 3. Проверка пропусков

In [7]:
df_cleaned = df.copy()

In [8]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).
df_cleaned.isna().sum()
df_cleaned = df_cleaned.dropna(subset=['emp_length'])
df_cleaned['rate'] = df_cleaned['rate'].fillna(df_cleaned['rate'].median())
df_cleaned.isna().sum()

id                0
age               0
income            0
home              0
emp_length        0
intent            0
amount            0
rate              0
status            0
percent_income    0
default           0
cred_length       0
dtype: int64


---

**Пояснения**:
- Удалим 9 строк с emp_length, 1,4% данных (малоекол-во, можно удалить).
- Заменим 66 строк по rate медианой (~12.2%), т.к медиана отражает «типичное значение» и не портит статистику.
 ---

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [9]:
df_cleaned[df_cleaned.duplicated()]

id   age  income  home  emp_length             intent  amount  rate  \
650  649  23.0   58800  RENT         7.0  DEBTCONSOLIDATION   20000  7.66   
651  649  23.0   58800  RENT         7.0  DEBTCONSOLIDATION   20000  7.66   

     status  percent_income default  cred_length  
650       1            0.34       N            3  
651       1            0.34       N            3

In [10]:
# удалите дубликаты, если они есть
df_cleaned = df_cleaned.drop_duplicates().reset_index(drop=True)
df_cleaned.duplicated().sum()

np.int64(0)

#### Проверка неявных дубликатов

In [11]:
unique_c = ['age', 'home', 'emp_length', 'intent', 'status', 'default', 'cred_length']
for i in unique_c:
  print(i,'\n', df_cleaned[i].unique())

age 
 [ 22.  21.  25.  23.  24.  26. 144. 123.]
home 
 ['RENT' 'OWN' 'MORTGAGE' 'RENET' 'OTHER']
emp_length 
 [123.   5.   1.   4.   8.   2.   6.   7.   0.   9.   3.  10.  11.]
intent 
 ['PERSONAL' 'EDUCATION' 'MEDICAL' 'VENTURE' 'HOMEIMPROVEMENT'
 'DEBTCONSOLIDATION']
status 
 [1 0]
default 
 ['Y' 'N' 'No']
cred_length 
 [3 2 4]


In [12]:
for i in unique_c:
  print(df[i].value_counts())

age
24.0     137
23.0     132
25.0     130
22.0     109
26.0      98
21.0      43
144.0      2
123.0      1
Name: count, dtype: int64
home
RENT        415
MORTGAGE    136
OWN          98
OTHER         2
RENET         1
Name: count, dtype: int64
emp_length
0.0      91
2.0      76
3.0      70
5.0      64
6.0      61
1.0      60
7.0      58
4.0      46
8.0      46
9.0      45
10.0     22
123.0     2
11.0      2
Name: count, dtype: int64
intent
EDUCATION            145
VENTURE              117
PERSONAL             111
MEDICAL              107
DEBTCONSOLIDATION    104
HOMEIMPROVEMENT       68
Name: count, dtype: int64
status
1    392
0    260
Name: count, dtype: int64
default
N     499
Y     150
No      3
Name: count, dtype: int64
cred_length
3    226
4    217
2    209
Name: count, dtype: int64


In [13]:
# удалите дубликаты, если они есть
df_cleaned['home'] = df_cleaned['home'].replace('RENET', 'RENT') # исправление типа жилья
df_cleaned['default'] = df_cleaned['default'].replace('No', 'N') # исправление категории дефолта

for i in unique_c:
  print(i,'\n', df_cleaned[i].unique())

age 
 [ 22.  21.  25.  23.  24.  26. 144. 123.]
home 
 ['RENT' 'OWN' 'MORTGAGE' 'OTHER']
emp_length 
 [123.   5.   1.   4.   8.   2.   6.   7.   0.   9.   3.  10.  11.]
intent 
 ['PERSONAL' 'EDUCATION' 'MEDICAL' 'VENTURE' 'HOMEIMPROVEMENT'
 'DEBTCONSOLIDATION']
status 
 [1 0]
default 
 ['Y' 'N']
cred_length 
 [3 2 4]


---

**Пояснения:**
Мы обнаружили и исправили неявные дубликаты:
- home - опечатка "RENET";
- default - смешение категорий (No и N);
 ---

### 5. Провека типов данных

In [14]:
df_cleaned.dtypes

df_cleaned['age'] = df_cleaned['age'].astype(int)
df_cleaned['emp_length'] = df_cleaned['emp_length'].astype(int)

df_cleaned['home'] = df_cleaned['home'].astype('category')
df_cleaned['intent'] = df_cleaned['intent'].astype('category')
df_cleaned['default'] = df_cleaned['default'].astype('category')
df_cleaned['status'] = df_cleaned['status'].astype('category')

In [15]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности.
df_cleaned.dtypes

id                   int64
age                  int64
income               int64
home              category
emp_length           int64
intent            category
amount               int64
rate               float64
status            category
percent_income     float64
default           category
cred_length          int64
dtype: object

---

**Пояснения**
Некоторые признаки были определены неверно:
age и emp_length были записаны как float64, должны быть целыми числами;
home, intent, default и status определены как object или int64, но являются категориальными.

Типы данных были приведены в соответствие

 ---

### 6. Группировка данных

#### Задание 1

*`Группировка - статус домовладения (home) по количеству дефолтов
(default) . Р`*

In [16]:
df_cleaned.groupby(['home', 'default'], observed=True)['id'].count()

home      default
MORTGAGE  N          102
          Y           28
OTHER     Y            2
OWN       N           83
          Y           13
RENT      N          307
          Y          106
Name: id, dtype: int64

истории дефолта.

**`Этот код группирует наших клиентов по двум признакам: по типу жилья (home) и по тому, был ли у них дефолт в прошлом (default). Потом он просто считает, сколько человек (записей id) попадает в каждую такую комбинацию. Мы видим, сколько людей с разным жилищным статусом портили свою кредитную историю в прошлом.`**

#### Задание 2

*`: Группировка - цель кредита (intent) по количеству статуса
домовладения (home). Создать датафрейм. Переименовать столбец с количеством
Лабораторная работа 1 Предварительный анализ данных 28
в “сount”. Отсортировать по убыванию столбца “count”`*

In [25]:
df_grouped = df_cleaned.groupby(['intent', 'home'], observed=True).size().reset_index(name='count')
df_grouped = df_grouped.sort_values(by='count', ascending=False).reset_index(drop=True)

(df_grouped)

intent      home  count
0           EDUCATION      RENT     94
1   DEBTCONSOLIDATION      RENT     72
2             MEDICAL      RENT     70
3             VENTURE      RENT     70
4            PERSONAL      RENT     68
5     HOMEIMPROVEMENT      RENT     39
6           EDUCATION  MORTGAGE     32
7             VENTURE  MORTGAGE     27
8             MEDICAL  MORTGAGE     22
9            PERSONAL       OWN     22
10            VENTURE       OWN     19
11  DEBTCONSOLIDATION  MORTGAGE     19
12           PERSONAL  MORTGAGE     18
13    HOMEIMPROVEMENT       OWN     16
14            MEDICAL       OWN     15
15          EDUCATION       OWN     15
16    HOMEIMPROVEMENT  MORTGAGE     12
17  DEBTCONSOLIDATION       OWN      9
18          EDUCATION     OTHER      1
19           PERSONAL     OTHER      1

**`Тут мы снова группируем данные, но теперь по цели кредита (intent) и типу жилья (home). Код считает количество заявок в каждой группе, а потом сортирует результат так, чтобы самая популярная комбинация цели и жилья была наверху списка. Мы видим, за какими именно целями кредита люди с разным жилищным статусом обращаются чаще всего.`**

#### Задание 3

*`Сводная таблица (pivot_table) - средний доход income для каждого
возраста. Отсортировать по убыванию. Округлить до двух знаков.`*

In [26]:
pivot_income = df_cleaned.pivot_table(
    values='income',
    index='age',
    aggfunc='mean'
).round(2).sort_index(ascending=False)

pivot_income = pivot_income.rename(columns={'income': 'avg_income'})

(pivot_income)


avg_income
age            
144   225000.00
123    80004.00
26    122793.74
25    107174.44
24     89020.24
23     85682.35
22     65794.82
21     43390.39

**`Здесь мы строим таблицу, где для каждого возраста (age) вычисляется средний доход (income) всех людей этого возраста в нашем наборе данных. Строки (возраст) отсортированы по убыванию, то есть наверху таблицы будут самые старшие клиенты. Мы видим, как в среднем меняется доход с возрастом. Обычно можно ожидать, что у более молодых людей (внизу списка) средний доход будет ниже`**

#### Задание 4

*`Сводная таблица (pivot_table) - средняя сумма кредита amount для
каждой цели (intent) - строки и возраста - столбцы . Отсортировать по
возрастанию intent. Округлить до двух знаков.`*

In [27]:
pivot_amount = df_cleaned.pivot_table(
    values='amount',
    index='intent',
    columns='age',
    aggfunc='mean',
    fill_value=0  
).round(2)

pivot_amount = pivot_amount.sort_index()

(pivot_amount)


/tmp/ipykernel_6328/3609056476.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_amount = df_cleaned.pivot_table(


age                     21        22        23        24        25        26   \
intent                                                                          
DEBTCONSOLIDATION  10583.33  18602.78  20291.67  21269.23  18603.33  21947.22   
EDUCATION          13500.00  19207.26  20288.28  20355.00  19326.04  20566.67   
HOMEIMPROVEMENT     4355.00  10216.67   4000.00  20095.37  21266.67  20038.46   
MEDICAL            10700.00  18073.21  17968.48  21119.64  18114.81  20410.00   
PERSONAL            3633.33  18522.50  19241.00  21205.00  21643.75  18588.24   
VENTURE             9407.69  16310.29  22810.94  19406.73  18360.42  18850.00   

age                    123     144  
intent                              
DEBTCONSOLIDATION      0.0     0.0  
EDUCATION          20400.0  6000.0  
HOMEIMPROVEMENT        0.0     0.0  
MEDICAL                0.0     0.0  
PERSONAL               0.0     0.0  
VENTURE                0.0  4800.0

**`Мы создаем "сводную таблицу". По строкам у нас цели кредита (intent), а по столбцам — возраст (age). На пересечении строки и столбца находится средняя сумма кредита (amount), которую запрашивают люди данного возраста для данной цели. По горизонтали (по возрасту): Как меняется средняя сумма кредита на одну и ту же цель (например, на образование) для разных возрастов. По вертикали (по цели): На какие цели в среднем берут самые крупные кредиты(можно заметить, что погашение кредита или оплата образования является лидером), а на какие — самые мелкие`**

### Вывод

В ходе работы был проведён предварительный анализ набора данных credit_risk.csv, содержащего информацию о 652 клиентах и их кредитах. Данные относятся к предметной области оценки кредитных рисков и включают показатели возраста, дохода, цели кредита, типа жилья, суммы займа, процентной ставки, длины кредитной истории и факта дефолта. На этапе предобработки были выявлены и устранены ошибки: некорректные значения возраста (123 и 144 года) и стажа работы (123 года, дробное значение 4.6), опечатка в типе жилья («RENET» заменена на «RENT»), а также неявный дубликат в колонке дефолта («No» объединено с «N»). Пропуски в числовых данных были заполнены медианными значениями. Типы данных приведены в соответствие: числовые признаки переведены в int64 или float64, категориальные признаки — в category.

Анализ группировок показал, что наибольшее количество клиентов берут кредиты, находясь в аренде жилья RENT. Среди них же наблюдается и наибольшее количество дефолтов (106 случаев). В то же время владельцы собственного жилья OWN составляют меньшинство выборки и демонстрируют наименьший уровень дефолтов (13 случаев). Ипотечники MORTGAGE занимают промежуточное положение. Таким образом, арендаторы являются наиболее уязвимой категорией заёмщиков.

В лабораторной работе осуществлялась работа с библиотекой pandas. Были изучены типы данных в библиотеке, индексация и др. На практике были произведены подготовка данных, устранение ошибок и дубликатов, обработка различных типов данных. Были изучены такие понятия как группировка и категоризация.

### Дополнительное задание

**`4. Создать столбец “Категория дохода (Income)” (с помощью категоризации).
Выделить минимум 3 категории (низкий, высокий, средний), фильтрацию для
уровня дохода выбрать самостоятельно, аргументировать выбор. Создать
группировку: средняя, минимальная, максимальная, медианная сумма кредита
(amount) по категории дохода.`**

In [36]:
df_cleaned['income'].describe()

count       641.000000
mean      90811.184087
std       69567.707227
min        9600.000000
25%       45000.000000
50%       70000.000000
75%      130000.000000
max      500000.000000
Name: income, dtype: float64

Я использовал метод describe(), для выбора границ доходов, т.к. он показывает нам основную статистику по столбцу с доходами - среднее значение, разброс данных, минимальный и максимальный доход, а также квартили распределения.

Мы выбрали границы по квартилям (Q1 = 45000, Q3 = 130000).\
Это логично, так как категории распределяются относительно большинства значений.

Сами границы:
Низкий доход <= 45000, \
Средний доход > 45000 и <= 130000, \
Крупный доход > 130000.

In [38]:
def income_category(income):
    if income <= 45000:
        return 'Низкий'
    elif income <= 130000:
        return 'Средний'
    else:
        return 'Высокий'

df_cleaned['income_cat'] = df_cleaned['income'].apply(income_category)

income_group = df_cleaned.groupby('income_cat')['amount'].agg(['mean', 'min', 'max', 'median']).round(2)
(income_group)

mean    min    max   median
income_cat                                 
Высокий     18941.46   1000  35000  20000.0
Низкий       8043.67   1000  25000   4537.5
Средний     24408.18  10000  35000  24250.0

Здесь я создал функцию income_category(), которая разбивает доходы на три категории: низкий, средний и высокий. Это нужно было для упрощения анализа вместо работы с непрерывными числовыми значениями. 

Для анализа использовал следующие методы:
- apply() использовался для применения этой функции ко всем строкам в столбце доходов, создавая новый категориальный столбец.
- groupby() для группировки данных по созданным категориям дохода
- метод agg() для расчета сразу нескольких статистик по сумме кредита в каждой группе: среднего, минимума, максимума и медианы.

Получаем следующую информацию:
Заемщики с высоким доходом берут в среднем кредиты на сумму около (~18941). Минимальная сумма (1000) может быть исключением, например, для образовательных целей.

Заемщики со средним доходом имеют самую высокую среднюю сумму кредита (~24408), а медиана равна 24250. Это самая активная группа, которая чаще всего берет крупные кредиты. Минимальная сумма (10000) довольно крупная, что говорит о том,что заемщики со средним доходом не берут мелкие суммы кредитов.

Заемщики с низким доходом берут значительно меньшие по сумме кредиты. Их средний кредит (~8043), а медиана — (~4537). Это логично, так как их доход не позволяет брать большие займы (что также видно по столбцу percent_income в исходных данных). При этом в этой группе тоже есть люди, берущие максимальные 25000, что может быть рискованно.


**`9. Создать столбец “Категория суммы кредита” - Amount (с помощью
категоризации). Выделить минимум 3 категории (маленькая, крупная, средняя),
фильтрацию для уровня суммы выбрать самостоятельно, аргументировать
выбор. Создать группировку: средняя, минимальная, максимальная,
медианная сумма кредита (amount) по цели кредита.`**

In [37]:
df_cleaned["amount"].describe()

count      641.000000
mean     18924.882995
std       9073.306111
min       1000.000000
25%      12000.000000
50%      22000.000000
75%      25000.000000
max      35000.000000
Name: amount, dtype: float64

Аналогично заданию 4 использовал метод desctibe.

Мы выбрали границы по квартилям (Q1 = 12000, Q3 = 25000).\
Это логично, так как категории распределяются относительно большинства значений.

Сами границы:
Маленькая сумма кредита <= 12000, \
Средняя сумма кредита > 12000 и <= 25000, \
Крупная сумма кредита > 25000.

In [39]:
def amount_category(amount):
    if amount <= 12000:
        return 'Маленькая'
    elif amount <= 25000:
        return 'Средняя'
    else:
        return 'Крупная'

df_cleaned['amount_cat'] = df_cleaned['amount'].apply(amount_category)

intent_group = df_cleaned.groupby('intent')['amount'].agg(['mean', 'min', 'max', 'median']).round(2)
(intent_group)

/tmp/ipykernel_6328/2144130189.py:11: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  intent_group = df_cleaned.groupby('intent')['amount'].agg(['mean', 'min', 'max', 'median']).round(2)


mean   min    max   median
intent                                           
DEBTCONSOLIDATION  19868.25  2275  35000  23800.0
EDUCATION          19266.73  1000  35000  21475.0
HOMEIMPROVEMENT    18047.01  1400  33000  21500.0
MEDICAL            18732.24  1200  35000  21250.0
PERSONAL           19467.20  1400  35000  24000.0
VENTURE            17868.32  1000  35000  20250.0

Аналогично доходам, я применил функцию amount_category() для разделения сумм кредитов на маленькие, средние и крупные. Это помогает анализировать не абсолютные значения, а более понятные категории.

Рейтинг целей по средней сумме кредита:
- DEBTCONSOLIDATION (консолидация долгов) - 19,868.25
- PERSONAL (личные цели) - 19,467.20
- EDUCATION (образование) - 19,266.73
- MEDICAL (медицина) - 18,732.24
- HOMEIMPROVEMENT (улучшение жилья) - 18,047.01
- VENTURE (бизнес) - 17,868.32

Консолидация долгов (DEBTCONSOLIDATION) демонстрирует наибольшую среднюю сумму кредита, что логично - для объединения нескольких существующих долгов обычно требуется значительная сумма. При этом медианное значение (23,800) даже выше среднего, что указывает на преобладание крупных кредитов в этой категории.

Личные цели (PERSONAL) занимают второе место по средней сумме с высоким медианным значением 24,000, что свидетельствует о популярности крупных кредитов для удовлетворения персональных потребностей.

Образовательные кредиты (EDUCATION) показывают широкий разброс - от небольших сумм в 1,000 (возможно, краткосрочные курсы) до максимальных 35,000 (вероятно, полное высшее образование).

Бизнес-кредиты (VENTURE) имеют наименьшую среднюю сумму, что может объясняться осторожностью как заемщиков (начинающих предпринимателей), так и кредиторов в финансировании рискованных ventures.

**`17. Создать столбец “Категория суммы кредита” - Amount (с помощью
категоризации). Выделить минимум 3 категории (маленькая, крупная, средняя),
фильтрацию для уровня суммы выбрать самостоятельно, аргументировать
выбор. Создать столбец “Категория дохода (Income)” (с помощью
категоризации). Выделить минимум 3 категории (низкий, высокий, средний),
фильтрацию для уровня дохода выбрать самостоятельно, аргументировать
выбор. Создать сводную таблицу: медианная сумма кредита по категории
суммы кредита и категории дохода.`**

In [40]:
pivot_table = df_cleaned.pivot_table(
    index='income_cat',
    columns='amount_cat',
    values='amount',
    aggfunc='median',
    fill_value=0
)

(pivot_table)

amount_cat  Крупная  Маленькая  Средняя
income_cat                             
Высокий     32400.0     9975.0  22000.0
Низкий          0.0     3625.0  20000.0
Средний     30000.0    10000.0  24000.0

Я использовал метод pivot_table() для создания сводной таблицы, которая показывает медианную сумму кредита на пересечении категорий дохода и размера кредита. Это позволяет увидеть, как эти две характеристики связаны между собой.

Здесь же я использовал уже созданные в заданиях 4 и 9 столбцы: income_cat и amount_cat. Здесь я построил сводную таблицу (pivot table), где строками стала категория дохода, столбцами — категория суммы кредита, а значениями — медианная сумма кредита.

Результаты:
1. Крупные кредиты:
Высокий доход: 32400 - самая высокая медианная сумма среди всех категорий
Средний доход: 30000 - значительная сумма, но ниже чем у высокодоходной группы
Низкий доход: 0 - полное отсутствие крупных кредитов

Это показывает четкую зависимость: чем выше доход, тем более крупные кредиты доступны заемщику. Банк практически не выдает крупные кредиты клиентам с низким доходом, что является разумной риск-политикой.

2. Средние кредиты:
Высокий доход: 22000
Средний доход: 24000 (максимум в этой категории)
Низкий доход: 20000

Интересно, что заемщики со средним доходом берут самые крупные кредиты в категории "средние". Возможно, это их "потолок" по кредитной нагрузке.

3. Маленькие кредиты:
Высокий доход: 9975
Средний доход: 10000
Низкий доход: 3625 (самый низкий показатель)

Здесь видна обратная тенденция - заемщики с низким доходом берут значительно меньшие суммы даже в категории "маленькие" кредиты. Это может быть связано с осторожностью банка или самого заемщика.

**`23. Отфильтровать набор данных. Выбрать топ 2 цели кредита по средней сумме
кредита + топ 2 статуса землевладения по медианной сумме кредита. Создать
сводные таблицы: цель кредита, статус домовладения - медианный, средний,
максимальный доход.`**

In [41]:
top_intent = df_cleaned.groupby('intent')['amount'].mean().nlargest(2).index.tolist()
print("Топ-2 цели по средней сумме кредита:", top_intent)

Топ-2 цели по средней сумме кредита: ['DEBTCONSOLIDATION', 'PERSONAL']


/tmp/ipykernel_6328/1398054357.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  top_intent = df_cleaned.groupby('intent')['amount'].mean().nlargest(2).index.tolist()


Я сделал группировку через groupby('intent'), чтобы разбить данные по целям кредита. Затем я выбрал столбец 'amount' и применил к нему функцию mean() для расчёта срел. суммы кредита для каждой цели. Потом я использовал метод nlargest(2), чтобы выбрать две цели с наибольшими средними суммами кредита и уже после обратился к .index полученного результата, чтобы извлечь названия этих топ-целей, и преобразовал их в список через .tolist() для удобства работы.

Наконец, я вывел результат через print(), чтобы показать, какие две цели кредита имеют самые высокие средние суммы.

In [42]:
top_home = df_cleaned.groupby('home')['amount'].median().nlargest(2).index.tolist()
print("Топ-2 статуса домовладения по медианной сумме кредита:", top_home)

Топ-2 статуса домовладения по медианной сумме кредита: ['RENT', 'MORTGAGE']


/tmp/ipykernel_6328/141924351.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  top_home = df_cleaned.groupby('home')['amount'].median().nlargest(2).index.tolist()


Аналогично анализу топ-2 цели по сред. сумме кредита было сделано и с топ-2 статусами домовладения.

Метод же nlargest() здесь использовался для нахождения двух самых популярных целей кредита по средней сумме и двух лучших статусов домовладения по медианной сумме.

In [1]:
filtered_df = df_cleaned[
    (df_cleaned['intent'].isin(top_intent)) &
    (df_cleaned['home'].isin(top_home))
]

NameError: name 'df_cleaned' is not defined

Метод isin() я применил для фильтрации данных, оставив только топовые категории по целям кредита и статусу домовладения.

In [48]:
pivot_other_income = filtered_df.pivot_table(
    index='intent',
    columns='home',
    values='income',
    aggfunc='mean'
)

(pivot_other_income)

/tmp/ipykernel_6328/2157712759.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_other_income = filtered_df.pivot_table(


home                    MORTGAGE          RENT
intent                                        
DEBTCONSOLIDATION  140517.894737  85176.986111
PERSONAL           107103.388889  86512.735294

In [49]:
pivot_other_income = filtered_df.pivot_table(
    index='intent',
    columns='home',
    values='income',
    aggfunc='max'
)

(pivot_other_income)

/tmp/ipykernel_6328/1735851521.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_other_income = filtered_df.pivot_table(


home               MORTGAGE      RENT
intent                               
DEBTCONSOLIDATION  500000.0  306000.0
PERSONAL           200000.0  280000.0

In [50]:
pivot_median_income = filtered_df.pivot_table(
    index='intent',
    columns='home',
    values='income',
    aggfunc='median'
)

(pivot_median_income)

/tmp/ipykernel_6328/1299566479.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot_median_income = filtered_df.pivot_table(


home               MORTGAGE     RENT
intent                              
DEBTCONSOLIDATION  162000.0  74034.0
PERSONAL           160000.0  77825.0

Здесь же я несколько раз использовал pivot_table() с разными агрегирующими функциями 'mean', 'max', 'median', чтобы посмотреть на доходы заемщиков под разными углами - средний доход, максимальный и медианный в разрезе целей кредита и типа жилья.

Я рассчитал среднюю сумму кредита для каждой цели и выбрал две цели с наибольшими значениями. Затем я рассчитал медианную сумму кредита для каждого статуса домовладения (home) и выбрал два статуса с наибольшими значениями. 

Я отфильтровал исходную таблицу, оставив только строки, которые попадают в топ-2 по цели И в топ-2 по статусу домовладения.

Анализ результов:

1. Явное преимущество ипотечников:
По всем показателям (кроме одного) заемщики с ипотекой имеют более высокий доход. Особенно заметна разница в категории DEBTCONSOLIDATION: средний доход ипотечников выше, чем у арендаторов. Это логично, так как наличие ипотеки обычно свидетельствует о более стабильном финансовом положении.

2. Интересное исключение:
В категории PERSONAL максимальный доход у арендаторов (280,000) превышает максимальный доход ипотечников (200,000). Это может указывать на то, что некоторые высокооплачиваемые люди сознательно выбирают аренду, возможно, из-за мобильности или инвестиционных стратегий

3. Разница в профиле заемщиков:
DEBTCONSOLIDATION: привлекает более обеспеченных заемщиков (особенно с ипотекой)
PERSONAL: более демократичная цель, с меньшим разрывом в доходах между группами

4. Медиана vs Среднее:
В категории DEBTCONSOLIDATION медиана превышает среднее значение, что указывает на смещение распределения в сторону более высоких доходов
В категории PERSONAL аналогичная тенденция, особенно выраженная у ипотечников

